blog/2024-04-24-secure-remix.mdx

```diff
@@ -0,0 +1,103 @@
+---
+slug: instant-user-management-and-sso-for-remix
+title: Securing Remix Apps with Keycloak
+author: Phase Two
+tags: [phase_two, tutorial, frameworks, remix, react, reactjs]
+---
+
+import PhaseTwoStarterInstructions from "/templates/blog/_phase_two_starter_instructions.mdx";
+import OIDC from "/templates/blog/_oidc_client_creation_client_auth.mdx";
+import NonAdminUser from "/templates/blog/_add_nonadmin_user.mdx";
+
+In this article we'll be using [Keycloak](https://www.keycloak.org/) to quickly augment an application with user management and SSO. We will demonstrate the integration by securing a page for logged-in users. This quickly provides a jump-off point to more complex integrations.
+
+:::info
+
+If you just want to skip to the code, visit the Phase Two [Remix example](https://github.com/p2-inc/examples/tree/main/frameworks/remix). We also have a plain [React example](https://phasetwo.io/blog/instant-user-managemenet-and-sso-for-reactjs).
+
+:::
+
+## Setting up a Keycloak Instance
+
+<details>
+  <summary>Instructions</summary>
+  <PhaseTwoStarterInstructions />
+</details>
+
+## Setting up an OIDC Client
+
+<details>
+  <summary>Instructions</summary>
+  <OIDC />
+</details>
+
+## Adding a Non-Admin User
+
+<details>
+  <summary>Instructions</summary>
+  <NonAdminUser />
+</details>
+
+## Setting up a Remix Project
+
+:::info
+We will use the Phase Two Remix example code here, but the logic could easily be applied to any existing application.
+:::
+
+1. Clone the Phase Two [example repo](https://github.com/p2-inc/examples/).
+2. Open the Remix [folder](https://github.com/p2-inc/examples/tree/main/frameworks/remix) within `/frameworks/remix`.
+3. Run `npm install` and then `npm run dev -- --port 3000`. This example leverages [remix-auth](https://github.com/sergiodxa/remix-auth) and [remix-keycloak](https://github.com/marsmars0x01/remix-keycloak) to provide HOC support.
+4. Open the `app/services/keycloak.server.ts` file. This is a server only file. We will be updating a few values from the prior section where we set up our OIDC client. Taking the values from the OIDC Client Config section, set those values in the code. While it is recommended to use Environment variables for the secret, for the purpose of this tutorial, paste in the **Client secret** from the OIDC client creation section for the value of `clientSecret`:
+
+  ```tsx
```

Comment:
Sorry, I wasn't quite clear

```
const kcConfig = {
 realm, 
 ... // other values
}
```

then in the function

```
export default new KeycloakStrategy(
   kcConfig,
   ({ profile }) => profile
  );
```

Reply:
The example https://github.com/p2-inc/examples/pull/4/files should be updated to match